File: go/strategy/strategy.go
package main

import "fmt"

type IFlyBehaviour interface {
	fly()
}

type LowFlyBehaviour struct {
}

func (behaviour *LowFlyBehaviour) fly() {
	fmt.Println("Flying low!")
}

type HighFlyBehaviour struct {
}

func (behaviour *HighFlyBehaviour) fly() {
	fmt.Println("Flying high!")
}

type Duck struct {
	flyBehaviour IFlyBehaviour
}

func newDuck(behaviour IFlyBehaviour) *Duck {
	duck := new(Duck)
	duck.flyBehaviour = behaviour
	return duck
}

func (duck *Duck) fly() {
	duck.flyBehaviour.fly()
}

type LittleDuck struct {
	duck *Duck
}

func (duck *LittleDuck) fly()  {
	duck.duck.fly()
}

func NewLittleDuck() *LittleDuck {
	duck := new(LittleDuck)
	duck.duck = newDuck(new(LowFlyBehaviour))
	return duck
}

type BigDuck struct {
	duck *Duck
}

func (duck *BigDuck) fly()  {
	duck.duck.fly()
}

func NewBigDuck() *BigDuck {
	duck := new(BigDuck)
	duck.duck = newDuck(new(HighFlyBehaviour))
	return duck
}

func main() {
	littleDuck := NewLittleDuck()
	littleDuck.fly()
	bigDuck := NewBigDuck()
	bigDuck.fly()
}

File: python/strategy/strategy.py
from abc import ABC


class IFlyBehaviour(ABC):
    def fly(self):
        pass


class LowFlyBehaviour(IFlyBehaviour):
    def fly(self):
        print("Flying low!")


class HighFlyBehaviour(IFlyBehaviour):
    def fly(self):
        print("Flying high!")


class Duck(ABC):
    def __init__(self, fly_behaviour: IFlyBehaviour):
        self._fly_behaviour = fly_behaviour

    def fly(self):
        self._fly_behaviour.fly()


class LittleDuck(Duck):
    def __init__(self):
        super(LittleDuck, self).__init__(LowFlyBehaviour())


class BigDuck(Duck):
    def __init__(self):
        super(BigDuck, self).__init__(HighFlyBehaviour())


if __name__ == '__main__':
    little_duck = LittleDuck()
    little_duck.fly()
    big_duck = BigDuck()
    big_duck.fly()
